fix: resolve useProfileRedirect race condition

manvinderjit opened the pull request:
## Summary

Resolve api race condition in the hook when running it in strict mode

## Tests

NA

## Additional information

Closes #224

manvinderjit committed:
fix: resolve useProfileRedirect race condition

manvinderjit committed:
fix: add default values for variables

manvinderjit committed:
fix: add profile status validity check

bcha92 commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -37,43 +46,79 @@ const getRedirectPathForStatus = (status: ProfileStatus): string => {
+export const getProfileStatus = async (signal?: AbortSignal): Promise<ProfileStatusResult> => {
+	let statusOkay = false;
+	let shouldRedirect = true;
+	let profileStatus;
```
It might be a good idea to declare an initial value of `null` for `profileStatus` as this may cause a type check error as `undefined` if left undeclared.

bcha92 commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -37,43 +46,79 @@ const getRedirectPathForStatus = (status: ProfileStatus): string => {
+			const data: ProfileStatusResponse = await response.json();
+			statusOkay = true;
+			shouldRedirect = true;
+			profileStatus = isValidStatus(data.status) ? data.status : null;
```
suggested change here since there's essentially no change in the initial `let` declarations if status code is not 200:
```ts
		if (response.status === 200) {
			const data: ProfileStatusResponse = await response.json();
			statusOkay = true;
			profileStatus = isValidStatus(data.status) ? data.status : null;
		}

        // you may need to add a new error throw condition to account for err.name === 'AbortError'
        throw new Error("Add your invalid response here");
```

bcha92 commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -37,43 +46,79 @@ const getRedirectPathForStatus = (status: ProfileStatus): string => {
+			statusOkay = false;
+			shouldRedirect = false;
+			profileStatus = null;
+		}
```
should omit the declarations if no change from initial values. The if/else output is exactly the same. Should we just do this instead if you're going to create another error handling issue for this later?
```ts
catch (err) {
    console.error('err', err); // <---- you should put in the `throw new Error()` value from the try block for logging in here
    shouldRedirect = false;
}
```

bcha92 reviewed (changes requested):
Just a few variable declaration cleanup needed that are not necessary if the value is unchanged.

bcha92 commented on src/hooks/useProfileRedirect.ts:
```diff
@@ -37,43 +46,79 @@
-export const getProfileStatus = async (): Promise<ProfileStatus | null> => {
+export const getProfileStatus = async (signal?: AbortSignal): Promise<ProfileStatusResult> => {
+	let statusOkay = false;
+	let shouldRedirect = true;
```
^pretty much this is the warning for the review

manvinderjit replied:
@bcha92 I just refactored the whole code. I realized I didn't need to make it this complex. 

Should have done better the first time. Apologies for the inconvenience.

manvinderjit committed:
refactor: rework component for cleaner code

helper function now returns the redirect path instead of the profile status.
add and return state to prevent component flashing.
rename variables for clarity.
